### ISSUE TITLE
Orbit Web

### ISSUE THREAD

**Issue by amit2103** [electronicarts/orbit#110]
Thanks for building a great framework.
I was using the orbit -chat sample. Now the sample says that we can start only one server in one machine because of port limitations. Now if I start two servers in two machines will they auto discover each other? How does failover work? It will be nice if you can simply help me know how are things like HA configured.

Lastly can we have a better doc for Orbit-Web stating its functionalities. The orbit doc for other components is fantastic, but a bit more of web will be better.

**Comment by JoeHegarty** [comment]
Hi, I'm glad you like the framework.
I don't have access to a pc right now as I'm travelling.

I'll take a closer look at your questions tomorrow. We did just merge some big changes so it's possible we broke something. I agree with your point about the docs and I'll see if we can get this added to our backlog.

**Comment by amit2103** [comment]
@JoeHegarty ...Thanks for the answer.  The example works for me...I am just confused as how failover HA, clustering works for Orbit web...how to configure them etc. Whats happens if I start two chat front end (jetty servers ) and place a load balancer infront of them ....I mean lets say I want to create an web socket cluster...how does it work...

**Comment by johnou** [comment]
@amit2103 clustering depends on your environment really but the default cluster peer implementation uses a jgroups based configuration, for AWS you might consider a tcp based config setting initial hosts via system property and a larger port range.

The port limitation is probably just Jetty being configured to bind to a single interface / port here https://github.com/electronicarts/orbit/blob/master/samples/chat/chat-frontend/src/main/resources/conf/orbit.yaml changing this would allow you to start multiple "front ends" but then you'd need some form of LB to sit in front of them.. HAProxy, ELB etc.

In a nutshell failover works by the "backends" (actor host JVMs) persisting the actor state and if that node dies the actor will be recreated in another node and the state will be loaded from the configured storage provider.

**Comment by amit2103** [comment]
@johnou  Thanks. I understood that the actors with the same state will be recreated. I was curious that whether the frontend itself can cluster itself like the backend...from your answer it looks so. 

If it's not much of a hassle consider adding some doc for the Web components.

**Comment by JoeHegarty** [comment]
Yes, Orbit is designed to be able to scale both backends and frontends seamlessly.
As Johno pointed out the only restriction right now is that the Jetty port is fixed so you can't start 2 instances of a frontend on the same box without tweaking that. Other than this one restriction with the orbit-web project, frontends and backends are identical in terms of node discovery and scaling.

I've added a task to our backlog to create docs for orbit-web.

Hope this answers your questions, let me know if it doesn't.

**Comment by amit2103** [comment]
@JoeHegarty :+1:  Thanks...Answers my queries perfectly. Just another minor question? What are the latencies against which you tested? 50ms 100ms ? I promise this is my last question :)

**Comment by JoeHegarty** [comment]
The latency/processing time you experience with Orbit is highly dependent on the use case.
Obviously the complexity of your actor implementation or API handlers is going to have a big impact.

For REST using JAX-RS with Orbit Web, you are going to see performance the same as any other application using Jetty.

Internally on Orbit, when calling actors, the overhead introduced by the framework is very small. The actual latency you experience on cross-node actor messaging is going to be determined primarily by the network layer (JGroups, Infinispan by default), the condition of the network, load at the time etc.

Unfortunately I don't have any hard data to share at the moment from any of our applications.

**Comment by amit2103** [comment]
@JoeHegarty  Thanks for the help. I will bench it in my network. You are right these things depend on my network. However the fact that Orbit creates minimal overhead is nice to know.
